src/completion.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 John Spray All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ffi::c_void;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+use crate::ceph::IoCtx;
+use crate::error::RadosResult;
+use crate::rados::{
+    rados_aio_cancel, rados_aio_create_completion2, rados_aio_get_return_value,
+    rados_aio_is_complete, rados_aio_release, rados_aio_wait_for_complete_and_cb,
+    rados_completion_t,
+};
+
+struct Completion {
+    inner: rados_completion_t,
+
+    // Box to provide a stable address for completion_complete callback
+    // Mutex to make Sync-safe for write from poll() vs read from completion_complete
+    waker: Box<std::sync::Mutex<Option<std::task::Waker>>>,
```

Comment:
I think this is the part that I'm not following. `std::task::Waker` is Sync safe according to the docs.

Reply:
My comment is mangling the terminology a bit -- I'm not really using Mutex to persuade rust that the type is Sync (as you say, it already is), but to ensure that our accesses within unsafe{} regions are sound.

I put the Mutex in here to get the memory fencing that come with it -- on architectures with weaker memory ordering, we might otherwise write the Some() from one core, and then have the callback woken up on another core where a load of the Option could return a stale None if we haven't used a synchronization primitive that does some memory fencing.

Reply:
Got it. That makes sense. Thanks